bubblewrap.c

```diff
@@ -1088,8 +1096,22 @@ setup_newroot (bool unshare_pid,
       if (op->dest &&
           (op->flags & NO_CREATE_DEST) == 0)
         {
+          unsigned parent_mode = 0755;
+
+          /* If we're creating a file as 0600 or 0700, also create
+           * parents as 0700, and if we're creating a file as 0640 or 0750,
+           * create parents as 0750. */
+
+          if (op->perms >= 0 &&
+              (op->perms & 0040) == 0)
```

Comment:
The comment talks about hardcoded permissions, but the logic here is "if g-w, then chmod(parent, g-wx)" right?  There are use cases in setuid binaries for execute-only.  In other words I think this might result in surprise for creating a file with a mode like `711` (`rwx--x--x`) because the parent directory would end up not executable.

Why not make the logic here match the description and explicitly test `if (op->perms & 0077 == 0)` or so?

Reply:
Previously, parent directories created for a file or directory were always 0755, but if we're creating the file with restrictive permissions, then it seems odd not to create the parent directories with similarly restrictive permissions, so I was aiming for least-astonishment. A caller can use something like `--perms 0751 --dir ANCESTOR` or `--chmod 0751 ANCESTOR` for each ancestor directory if they want to be explicit (in practice this is probably only useful in the somewhat rarer case where you are starting a container that will be used by multiple uids).

What I was aiming for here was that for each of "group" and "other", if the file is not going to be readable by that entity, then make any newly-created parent directories also not readable by that entity. I'm treating "readable" (+r) and "traversable" (directories with +x) as basically equivalent for these purposes, because for directories you nearly always want them to be the same.

You're right that this will be odd with execute-only or write-only files, so I might be better with:

```diff
 if (op->perms >= 0 &&
-    (op->perms & 0040) == 0)
+    (op->perms & 0070) == 0) 
   parent_mode &= ~0050;
```

and the same for the "other".

I think I want to continue to do this separately for "group" and "other", though: if I'm doing

```
--perms 0750 --dir /one/two/three
```

in an otherwise empty filesystem, then it seems less astonishing to me to create `/one` and `/one/two` with 0750 permissions rather than 0755.

Similarly people occasionally use groups as "anticapabilities", so if I do

```
--perms 0707 --dir /one/two/three
```

it seems less surprising to create `/one` and `/two` with 0705 permissions ("everyone can read/traverse, except the owning group" rather than 0755.

Reply:
If you like that logic better, perhaps a better comment would be:

```
          /* If we're creating a file that is inaccessible by the owning group,
           * try to achieve least-astonishment by creating parents also
           * inaccessible by that group. */
          ... code involving the 0070 bits...

          /* The same, but for users other than the owner and group. */
          ... code involving the 0007 bits...
```

Reply:
> but the logic here is "if g-w, then chmod(parent, g-wx)" right?

No, the logic as currently implemented is:

```
always create parent u+rwx,g-w,o-w

if g+r
    create parent g+rx
else
    create parent g-rx

if o+r
    create parent o+rx
else
    create parent o-rx
```

and what I'm now suggesting in response to your review is changing that to:

```
always create parent u+rwx,g-w,o-w

if g+r or g+w or g+x
    create parent g+rx
else
    create parent g-rx

if o+r or o+w or o+x
    create parent o+rx
else
    create parent o-rx
```